src/ipc.rs

```diff
@@ -589,32 +600,52 @@ impl Serialize for IpcSharedMemory {
     where
         S: Serializer,
     {
-        let index = OS_IPC_SHARED_MEMORY_REGIONS_FOR_SERIALIZATION.with(
-            |os_ipc_shared_memory_regions_for_serialization| {
-                let mut os_ipc_shared_memory_regions_for_serialization =
-                    os_ipc_shared_memory_regions_for_serialization.borrow_mut();
-                let index = os_ipc_shared_memory_regions_for_serialization.len();
-                os_ipc_shared_memory_regions_for_serialization.push(self.os_shared_memory.clone());
-                index
-            },
-        );
-        index.serialize(serializer)
+        if let Some(os_shared_memory) = &self.os_shared_memory {
+            let index = OS_IPC_SHARED_MEMORY_REGIONS_FOR_SERIALIZATION.with(
+                |os_ipc_shared_memory_regions_for_serialization| {
+                    let mut os_ipc_shared_memory_regions_for_serialization =
+                        os_ipc_shared_memory_regions_for_serialization.borrow_mut();
+                    let index = os_ipc_shared_memory_regions_for_serialization.len();
+                    os_ipc_shared_memory_regions_for_serialization.push(os_shared_memory.clone());
+                    index
+                },
+            );
+            debug_assert!(index < usize::MAX);
+            index
+        } else {
+            usize::MAX
+        }
+        .serialize(serializer)
     }
 }
 
 impl IpcSharedMemory {
+    const fn empty() -> Self {
+        Self {
+            os_shared_memory: None,
+        }
+    }
+
     /// Create shared memory initialized with the bytes provided.
     pub fn from_bytes(bytes: &[u8]) -> IpcSharedMemory {
-        IpcSharedMemory {
-            os_shared_memory: OsIpcSharedMemory::from_bytes(bytes),
+        if bytes.is_empty() {
+            IpcSharedMemory::empty()
+        } else {
+            IpcSharedMemory {
+                os_shared_memory: Some(OsIpcSharedMemory::from_bytes(bytes)),
+            }
```

Comment:
Could this be made even simpler by having this return `Option<IpcSharedMemory>`, much like `ptr::NonNull`? Then calling code would explicitly need to handle when the shared memory is empty. It feels like it should be an error to try to create an empty shared memory segment -- as it's kinda useless.

Reply:
I think that would made uses more verbose (most of the time there would be Some value so a lot of None values would be unreachable) and I feel this would complicate things for consumers. There is also no way in rust to tell that slice is at least length of 1, so consumers would still need to handle len=0 cases, even if they get Some. Also this would be breaking change and I tried to avoided that.

Empty buffer also has meaning (receiver was woken with empty message).

Reply:
The thing is that this API needs to change anyway. Creating a shared memory segment can fail (see the manpage for `shm_open` for instance). ipc-channels is pretending like it cannot fail and it will panic if a failure does occur. When creating IPC channels, this reality is acknowledged, because `channel()` returns a `Result`. Not doing the same here feels like a bit of a lie. Many programs will want to handle failure gracefully.

Reply:
Indeed, but I think better error handling should be done as in separate PR. This PR just makes current API works as expected.